pkg/clusteragent/appsec/injector.go

```diff
@@ -0,0 +1,356 @@
+	detected := make(map[appsecconfig.ProxyType]struct{})
+	for proxy, detector := range proxyDetectionMap {
+		found, err := detector(ctx, cl.DynamicCl)
+		if err != nil {
+			logger.Warnf("error detecting proxy %s in cluster: %s", proxy, err)
+			continue
+		}
+		if found {
+			detected[proxy] = struct{}{}
+		}
+	}
+
+	return detected, nil
+}
+
+type securityInjector struct {
+	ctx                   context.Context
+	cancel                context.CancelFunc
+	k8sClient             dynamic.Interface
+	logger                log.Component
+	config                appsecconfig.Config
+	recorder              record.EventRecorder
+	leaderElectionEnabled bool
+	baseBackoff           time.Duration
+	maxBackoff            time.Duration
+}
+
+// NewLanguagePatcher initializes and returns a new patcher with a dynamic k8s client
+func newSecurityInjector(ctx context.Context, logger log.Component, datadogConfig config.Component) *securityInjector {
+	config := appsecconfig.FromComponent(datadogConfig)
+	if !config.Injection.Enabled && !config.Product.Enabled {
+		logger.Info("Appsec proxy injection is disabled")
+		return nil
+	}
+
+	// Get API client to create event recorder
+	apiClient, err := apiserver.GetAPIClient()
+	if err != nil {
+		logger.Errorf("Failed to get API client for event recorder: %v", err)
+		return nil
+	}
+
+	if config.AutoDetect {
+		detectedProxies, err := detectProxiesInCluster(ctx, apiClient, logger)
+		if err != nil {
+			logger.Warnf("error detecting proxies in cluster: %s", err)
+		}
+		maps.Copy(config.Proxies, detectedProxies)
+	}
+
+	if len(config.Proxies) == 0 {
+		logger.Info("No appsec proxies enabled for injection")
+		return nil
+	}
+
+	// Create event recorder
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartRecordingToSink(&corev1.EventSinkImpl{
+		Interface: apiClient.Cl.CoreV1().Events(config.Processor.Namespace),
+	})
+	eventRecorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{
+		Component: "datadog-appsec-injector",
+	})
+
+	ctx, cancel := context.WithCancel(ctx)
+	return &securityInjector{
+		ctx:                   ctx,
+		cancel:                cancel,
+		k8sClient:             apiClient.DynamicCl,
+		logger:                logger,
+		config:                config,
+		recorder:              eventRecorder,
+		leaderElectionEnabled: datadogConfig.GetBool("leader_election"),
+		baseBackoff:           datadogConfig.GetDuration("cluster_agent.appsec.injector.base_backoff"),
+		maxBackoff:            datadogConfig.GetDuration("cluster_agent.appsec.injector.max_backoff"),
+	}
+}
+
+type workItemType int
+
+const (
+	_ workItemType = iota
+	workItemAdded
+	workItemModified
+	workItemDeleted
+)
+
+func (sub workItemType) String() string {
+	switch sub {
+	case workItemAdded:
+		return "added"
+	case workItemModified:
+		return "modified"
+	case workItemDeleted:
+		return "deleted"
+	default:
+		return "unknown"
+	}
+}
+
+type workItem struct {
```

Comment:
It might be possible to simplify this struct. This is related to my other comment in `pkg/clusteragent/appsec/config/pattern.go`